### PR TITLE
Implement incremental bulk execution

### PR DESCRIPTION
**Pull request by Tim-Brooks**
This commit adds a mechanism of splitting bulk requests when a
coordinating nodes is under memory pressure. In order to do that it
adds a new mechanism of http stream handling. And the proper framework
to use this with bulk requests.

### CODE REVIEW

**Review by nielsbauman — COMMENTED** [review]
The changes in `BulkOperation` LGTM. I do think it would be worth adding one test that ensures that short-circuited shard requests also end up in the failure store - maybe in `IncrementalBulkRestIT.java` or `IncrementalBulkIT.java` perhaps.

File: server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java [hunk 1]
```diff
@@ -117,12 +119,28 @@ public final void handleRequest(RestRequest request, RestChannel channel, NodeCl
...
-            if (request.hasContent() && request.isContentConsumed() == false) {
+            if (request.hasContent() && (request.isContentConsumed() == false && request.isFullContent())) {
                 throw new IllegalArgumentException(
                     "request [" + request.method() + " " + request.path() + "] does not support having a body"
                 );
             }
 
+            if (request.isStreamedContent()) {
+                assert action instanceof RequestBodyChunkConsumer;
+                var chunkConsumer = (RequestBodyChunkConsumer) action;
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java:130]: I think this is where we'd fail if there is a discrepancy between the handlers and the `decider` filter in `Netty4HttpAggregator`. This is fine for now (for the merge), and possibly also for our initial roll-out plans.

I do wonder though if we could have a simple aggregating handler here that protects us against a discrepancy, in particular before we enable this for regular releases (or perhaps before that even)? Would it be feasible to always do the aggregation here even (when the mode is enabled), assuming larger requests are handled by proper handlers?

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpAggregator.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+    public Netty4HttpAggregator(int maxContentLength, Predicate<HttpPreRequest> decider) {
+        super(maxContentLength);
+        this.decider = decider;
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        assert msg instanceof HttpObject;
+        if (msg instanceof HttpRequest request) {
+            var preReq = HttpHeadersAuthenticatorUtils.asHttpPreRequest(request);
```
Comment (henningandersen) [inline modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpAggregator.java:48]: I find this somewhat unnecessary, could the decider not work on the uri directly instead or even the `HttpRequest` (the decider is quite intimate anyway). Happy to leave as is for now ofc.

**Review by henningandersen — APPROVED** [review]
LGTM.

**Comment by henningandersen** [comment]
> I do think it would be worth adding one test that ensures that short-circuited shard requests also end up in the failure store - maybe in IncrementalBulkRestIT.java or IncrementalBulkIT.java perhaps.

Thanks, that makes sense. We should tackle that as a follow-up though, given that the incremental bulk handling is off by default and I'd prefer to get this in and start some early testing on it.

File: server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java [hunk 3]
```diff
@@ -429,90 +436,102 @@ private void discardRedirectsAndFinish(Exception exception) {
     }
 
     private void executeBulkShardRequest(BulkShardRequest bulkShardRequest, Releasable releaseOnFinish) {
-        client.executeLocally(TransportShardBulkAction.TYPE, bulkShardRequest, new ActionListener<>() {
+        ShardId shardId = bulkShardRequest.shardId();
 
-            // Lazily get the cluster state to avoid keeping it around longer than it is needed
-            private ClusterState clusterState = null;
+        // Short circuit the shark level request with the existing shard failure.
```
Comment (jbaiera) [inline server/src/main/java/org/elasticsearch/action/bulk/BulkOperation.java:441]:
```suggestion
        // Short circuit the shard level request with the existing shard failure.
```
🦈

**Review by jbaiera — APPROVED** [review]
BulkOperation changes look like they make sense to me! I second the ask for a test, but it's fine if that comes as a follow up.

File: server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java (hunk 1 above)
Comment (mhl-b) [reply server/src/main/java/org/elasticsearch/rest/BaseRestHandler.java:130]: It's indeed a temporary measure to do minimal change in http/rest layer to allow partial content. Current intention is to remove HttpAggregator and do aggregation in RestHandler(Base?). I drafted it before in #112120 and planning to revisit once we merge to main. There are several options how to do that, I will prepare a design doc.

File: modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpAggregator.java (hunk 2 above)
Comment (Tim-Brooks) [reply modules/transport-netty4/src/main/java/org/elasticsearch/http/netty4/Netty4HttpAggregator.java:48]: Yes. IMO we definitely have room for improvement here. There is a ticket tracking improved handling in this area.

**Review by mhl-b — APPROVED** [review]
LGTM

**Pull request merged**